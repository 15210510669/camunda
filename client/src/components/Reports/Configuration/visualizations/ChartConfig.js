import React from 'react';
import ShowInstanceCount from './subComponents/ShowInstanceCount';
import {ColorPicker, Switch, LabeledInput} from 'components';
import RelativeAbsoluteSelection from './subComponents/RelativeAbsoluteSelection';
import ChartTargetInput from './subComponents/ChartTargetInput';
import './ChartConfig.scss';

export default function ChartConfig({configuration, onChange, report}) {
  const enabledTarget = configuration.targetValue ? configuration.targetValue.active : false;
  const defaultValues = {target: '', dateFormat: '', isBelow: false};

  const tooltipOptions = (
    <fieldset className="tooltipOptions">
      <legend>Tooltips options</legend>
      <RelativeAbsoluteSelection
        relativeDisabled={report.data.view.property !== 'frequency'}
        configuration={configuration}
        onChange={onChange}
      />
    </fieldset>
  );

  let typeSpecificComponents = [];
  switch (report.data.visualization) {
    case 'line':
      typeSpecificComponents.push(
        <div key="1" className="entry">
          <Switch
            checked={!configuration.pointMarkers}
            onChange={({target: {checked}}) => onChange('pointMarkers', !checked)}
          />
          Disable point markers
        </div>
      );
    /* falls through */
    case 'bar':
      typeSpecificComponents.push(
        <div className="fieldsetsContainer" key="2">
          <fieldset className="ColorSection">
            <legend>Select visualization color</legend>
            <ColorPicker selectedColor={configuration.color} onChange={onChange} />
          </fieldset>
          {tooltipOptions}
          <fieldset className="axisConfig">
            <legend>Axis names</legend>
            <LabeledInput
              label="x-axis"
              type="text"
              value={configuration.xLabel}
              onChange={({target: {value}}) => onChange('xLabel', value)}
            />
            <LabeledInput
              label="y-axis"
              type="text"
              value={configuration.yLabel}
              onChange={({target: {value}}) => onChange('yLabel', value)}
            />
          </fieldset>
          <fieldset className="goalLine" disabled={!enabledTarget}>
            <legend>
              <Switch
                checked={enabledTarget}
                onChange={({target: {checked}}) =>
                  onChange('targetValue', {
                    values: defaultValues,
                    ...configuration.targetValue,
                    active: checked
                  })
                }
              />
              Goal
            </legend>
            <ChartTargetInput {...{configuration, onChange, report}} />
          </fieldset>
        </div>
      );
      break;
    case 'pie':
      typeSpecificComponents.push(
        <div className="fieldsetsContainer" key="2">
          {tooltipOptions}
        </div>
      );
      break;
    default:
      typeSpecificComponents = [];
  }
  return (
    <>
      <ShowInstanceCount configuration={configuration} onChange={onChange} />
      {typeSpecificComponents}
    </>
  );
}

ChartConfig.defaults = {
  showInstanceCount: false,
  color: '#1991c8',
  pointMarkers: true,
  hideRelativeValue: false,
  hideAbsoluteValue: false,
  xLabel: '',
  yLabel: ''
};
