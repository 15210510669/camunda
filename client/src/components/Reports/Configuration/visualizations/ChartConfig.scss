.fieldsetsContainer fieldset {
  border: 1px solid #ddd;
  border-radius: 8px;
  margin: 5px 0;

  &.goalLine[disabled] .ChartTargetInput,
  &.tooltipOptions[disabled] {
    opacity: 0.4;
    pointer-events: none;
  }
}
