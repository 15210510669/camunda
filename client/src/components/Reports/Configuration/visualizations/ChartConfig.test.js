import React from 'react';
import {shallow} from 'enzyme';
import ChartConfig from './ChartConfig';

const getReportByVis = visualization => ({data: {visualization, view: {property: 'frequency'}}});

const lineReport = getReportByVis('line');
const barReport = getReportByVis('bar');
const pieReport = getReportByVis('pie');

const configuration = {
  showInstanceCount: false,
  color: '#1991c8',
  pointMarkers: true,
  hideRelativeValue: false,
  hideAbsoluteValue: false,
  xLabel: '',
  yLabel: '',
  targetValue: {active: false}
};

it('it should display correct configuration for linechart', () => {
  const node = shallow(<ChartConfig {...{report: lineReport, configuration}} />);
  expect(node).toMatchSnapshot();
});

it('it should display correct configuration for barchart', () => {
  const node = shallow(<ChartConfig {...{report: barReport, configuration}} />);
  expect(node).toMatchSnapshot();
});

it('it should display correct configuration for piechart', () => {
  const node = shallow(<ChartConfig {...{report: pieReport, configuration}} />);
  expect(node).toMatchSnapshot();
});

it('should invok onchange when changing switch for the point markers on line chart', () => {
  const spy = jest.fn();
  const node = shallow(<ChartConfig {...{report: lineReport, configuration}} onChange={spy} />);

  node
    .find('Switch')
    .first()
    .simulate('change', {target: {checked: true}});

  expect(spy).toHaveBeenCalledWith('pointMarkers', false);
});

it('should invok onchange when changing xlabel/ylabel input on bar chart', () => {
  const spy = jest.fn();
  const node = shallow(<ChartConfig {...{report: barReport, configuration}} onChange={spy} />);

  node
    .find('LabeledInput')
    .first()
    .simulate('change', {target: {value: 'testLabel'}});

  expect(spy).toHaveBeenCalledWith('xLabel', 'testLabel');
});

it('should invok onchange when enabling target value switch', () => {
  const spy = jest.fn();
  const node = shallow(<ChartConfig {...{report: lineReport, configuration}} onChange={spy} />);

  node.find('.goalLine Switch').simulate('change', {target: {checked: true}});

  expect(spy).toHaveBeenCalledWith('targetValue', {
    active: true,
    values: {dateFormat: '', isBelow: false, target: ''}
  });
});
