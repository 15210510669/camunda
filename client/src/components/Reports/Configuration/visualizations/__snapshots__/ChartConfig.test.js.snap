// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should display correct configuration for barchart 1`] = `
<Fragment>
  <ShowInstanceCount
    configuration={
      Object {
        "color": "#1991c8",
        "hideAbsoluteValue": false,
        "hideRelativeValue": false,
        "pointMarkers": true,
        "showInstanceCount": false,
        "targetValue": Object {
          "active": false,
        },
        "xLabel": "",
        "yLabel": "",
      }
    }
  />
  <div
    className="fieldsetsContainer"
    key="2"
  >
    <fieldset
      className="ColorSection"
    >
      <legend>
        Select visualization color
      </legend>
      <ColorPicker
        colors={
          Array [
            "#B80000",
            "#DB3E00",
            "#FCCB00",
            "#008B02",
            "#006B76",
            "#1991c8",
            "#004DCF",
            "#5300EB",
            "#EB9694",
            "#FAD0C3",
            "#FEF3BD",
            "#C1E1C5",
            "#BEDADC",
            "#C4DEF6",
            "#BED3F3",
            "#D4C4FB",
          ]
        }
        selectedColor="#1991c8"
      />
    </fieldset>
    <fieldset
      className="tooltipOptions"
    >
      <legend>
        Tooltips options
      </legend>
      <RelativeAbsoluteSelection
        configuration={
          Object {
            "color": "#1991c8",
            "hideAbsoluteValue": false,
            "hideRelativeValue": false,
            "pointMarkers": true,
            "showInstanceCount": false,
            "targetValue": Object {
              "active": false,
            },
            "xLabel": "",
            "yLabel": "",
          }
        }
        relativeDisabled={false}
      />
    </fieldset>
    <fieldset
      className="axisConfig"
    >
      <legend>
        Axis names
      </legend>
      <LabeledInput
        label="x-axis"
        onChange={[Function]}
        type="text"
        value=""
      />
      <LabeledInput
        label="y-axis"
        onChange={[Function]}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="goalLine"
      disabled={true}
    >
      <legend>
        <Switch
          checked={false}
          onChange={[Function]}
        />
        Goal
      </legend>
      <ChartTargetInput
        configuration={
          Object {
            "color": "#1991c8",
            "hideAbsoluteValue": false,
            "hideRelativeValue": false,
            "pointMarkers": true,
            "showInstanceCount": false,
            "targetValue": Object {
              "active": false,
            },
            "xLabel": "",
            "yLabel": "",
          }
        }
        report={
          Object {
            "data": Object {
              "view": Object {
                "property": "frequency",
              },
              "visualization": "bar",
            },
          }
        }
      />
    </fieldset>
  </div>
</Fragment>
`;

exports[`it should display correct configuration for linechart 1`] = `
<Fragment>
  <ShowInstanceCount
    configuration={
      Object {
        "color": "#1991c8",
        "hideAbsoluteValue": false,
        "hideRelativeValue": false,
        "pointMarkers": true,
        "showInstanceCount": false,
        "targetValue": Object {
          "active": false,
        },
        "xLabel": "",
        "yLabel": "",
      }
    }
  />
  <div
    className="entry"
    key="1"
  >
    <Switch
      checked={false}
      onChange={[Function]}
    />
    Disable point markers
  </div>
  <div
    className="fieldsetsContainer"
    key="2"
  >
    <fieldset
      className="ColorSection"
    >
      <legend>
        Select visualization color
      </legend>
      <ColorPicker
        colors={
          Array [
            "#B80000",
            "#DB3E00",
            "#FCCB00",
            "#008B02",
            "#006B76",
            "#1991c8",
            "#004DCF",
            "#5300EB",
            "#EB9694",
            "#FAD0C3",
            "#FEF3BD",
            "#C1E1C5",
            "#BEDADC",
            "#C4DEF6",
            "#BED3F3",
            "#D4C4FB",
          ]
        }
        selectedColor="#1991c8"
      />
    </fieldset>
    <fieldset
      className="tooltipOptions"
    >
      <legend>
        Tooltips options
      </legend>
      <RelativeAbsoluteSelection
        configuration={
          Object {
            "color": "#1991c8",
            "hideAbsoluteValue": false,
            "hideRelativeValue": false,
            "pointMarkers": true,
            "showInstanceCount": false,
            "targetValue": Object {
              "active": false,
            },
            "xLabel": "",
            "yLabel": "",
          }
        }
        relativeDisabled={false}
      />
    </fieldset>
    <fieldset
      className="axisConfig"
    >
      <legend>
        Axis names
      </legend>
      <LabeledInput
        label="x-axis"
        onChange={[Function]}
        type="text"
        value=""
      />
      <LabeledInput
        label="y-axis"
        onChange={[Function]}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="goalLine"
      disabled={true}
    >
      <legend>
        <Switch
          checked={false}
          onChange={[Function]}
        />
        Goal
      </legend>
      <ChartTargetInput
        configuration={
          Object {
            "color": "#1991c8",
            "hideAbsoluteValue": false,
            "hideRelativeValue": false,
            "pointMarkers": true,
            "showInstanceCount": false,
            "targetValue": Object {
              "active": false,
            },
            "xLabel": "",
            "yLabel": "",
          }
        }
        report={
          Object {
            "data": Object {
              "view": Object {
                "property": "frequency",
              },
              "visualization": "line",
            },
          }
        }
      />
    </fieldset>
  </div>
</Fragment>
`;

exports[`it should display correct configuration for piechart 1`] = `
<Fragment>
  <ShowInstanceCount
    configuration={
      Object {
        "color": "#1991c8",
        "hideAbsoluteValue": false,
        "hideRelativeValue": false,
        "pointMarkers": true,
        "showInstanceCount": false,
        "targetValue": Object {
          "active": false,
        },
        "xLabel": "",
        "yLabel": "",
      }
    }
  />
  <div
    className="fieldsetsContainer"
    key="2"
  >
    <fieldset
      className="tooltipOptions"
    >
      <legend>
        Tooltips options
      </legend>
      <RelativeAbsoluteSelection
        configuration={
          Object {
            "color": "#1991c8",
            "hideAbsoluteValue": false,
            "hideRelativeValue": false,
            "pointMarkers": true,
            "showInstanceCount": false,
            "targetValue": Object {
              "active": false,
            },
            "xLabel": "",
            "yLabel": "",
          }
        }
        relativeDisabled={false}
      />
    </fieldset>
  </div>
</Fragment>
`;
